services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/hakemus/HakemusUpdateRequest.kt

```diff
@@ -48,6 +55,18 @@ sealed interface HakemusUpdateRequest {
     fun toApplicationData(baseData: ApplicationData): ApplicationData
 
     fun customersByRole(): Map<ApplicationContactType, CustomerWithContactsRequest?>
+
+    /** Validate draft application. Checks only fields that have some actual data. */
+    fun validateForErrors(): ValidationResult =
```

Comment:
The validation would be better in a separate file like they were. The data classes should be lean, without too much functionality. Simulating the polymorphism needs a bit of boilerplate, but I think the separation of concerns is worth it.

The entrypoint can decide which type's method it will call:
```
fun HakemusUpdateRequest.validateForErrors(): ValidationResult =
    when (this) {
        is JohtoselvityshakemusUpdateRequest -> validateForErrors()
        is KaivuilmoitusUpdateRequest -> validateForErrors()
    }
```

The base class can have the common part, with a separate name:
```
fun HakemusUpdateRequest.validateCommonFieldsForErrors(): ValidationResult =
    validate { notJustWhitespace(name, "name") }
...
```

The implementations can call the common validation thanks to the distinct name:
```
fun JohtoselvityshakemusUpdateRequest.validateForErrors(): ValidationResult =
    this.validateCommonFieldsForErrors()
        .whenNotNull(postalAddress) { it.validateForErrors("postalAddress") }
...
```

Reply:
Or the common validation and the type decision can be done in the same function. This way the subtypes don't have to call the common validation.

You could even inline the `validateCommonFieldsForErrors()` to this function, but that might not be very readable.
```kotlin
fun HakemusUpdateRequest.validateForErrors(): ValidationResult =
    validateCommonFieldsForErrors().and {
        when (this) {
            is JohtoselvityshakemusUpdateRequest -> validateForErrors()
            is KaivuilmoitusUpdateRequest -> validateForErrors()
        }
    }
```

Reply:
Maybe you're right. I thought that while these "request" classes are only for controllers to receive (valid or invalid) data from the client, the validation could be inside the request class itself. But I will move it to the validator class.